hera_pspec/uvwindow.py

```diff
@@ -0,0 +1,1320 @@
+            # array are included and there is no over-sampling
+            kpara_bins = self.get_kpara_bins(self.freq_array,
+                                             self.little_h, self.cosmo)
+        else:
+            self.check_kunits(kpara_bins)
+        kpara_bins = np.array(kpara_bins.value)
+        nbins_kpara = kpara_bins.size
+
+        # COMPUTE THE WINDOW FUNCTIONS
+
+        # get cylindrical window functions for each baseline length considered
+        # as a function of (kperp, kpara)
+        # the kperp and kpara bins are given as global parameters
+        cyl_wf = np.zeros((nbls, self.Nfreqs, nbins_kperp, nbins_kpara))
+        for ib in range(nbls):
+            cyl_wf[ib, :, :, :] = self.get_cylindrical_wf(bl_lens[ib],
+                                                          kperp_bins*self.kunits,
+                                                          kpara_bins*self.kunits)*bl_weights[ib]
+
+        # Write file
+        with h5py.File(filepath, 'w') as f:
+            group = f.create_group('parameters')
+            group.create_dataset('avg_nu',shape=(1,),data=self.avg_nu,dtype=float)
+            group.create_dataset('avg_z',shape=(1,),data=self.avg_z,dtype=float)
+            group.create_dataset('nfreqs',shape=(1,),data=self.Nfreqs,dtype=int)
+            group.create_dataset('little_h',shape=(1,),data=self.little_h,dtype=bool)
+            group.create_dataset('polpair',shape=(1,),data=uvputils.polpair_tuple2int(self.pols),dtype=int)
+            ascii_taper = [n.encode("ascii", "ignore") for n in self.taper]
+            group.create_dataset('taper',(len(ascii_taper),),'S10',ascii_taper)
+            dset = f.create_group('window_functions')
+            dset.create_dataset('kperp_bins',shape=(nbins_kperp,),data=kperp_bins,dtype=float)
+            dset.create_dataset('kpara_bins',shape=(nbins_kpara,),data=kpara_bins,dtype=float)
+            dset.create_dataset('bl_lens',shape=(nbls,),data=bl_lens,dtype=float)
+            dset.create_dataset('bl_weights',shape=(nbls,),data=bl_weights,dtype=float)
+            dset.create_dataset('dly_array',shape=(self.Nfreqs,),data=self.dly_array,dtype=float)
+            dset.create_dataset('cyl_wf',shape=(nbls,self.Nfreqs,nbins_kperp,nbins_kpara),data=cyl_wf,dtype=float)
+            
+    def check_kunits(self, karray):
+        """
+        Check unit consistency between k's throughout code.
+
+        Parameters
+        ----------
+        karray : array-like
+            Array of Fourier modes.
+        """
+        try:
+            karray.unit
+        except AttributeError:
+            raise AttributeError('Feed k array with units (astropy.units).')
+        assert self.kunits.is_equivalent(karray.unit),\
+            "k array units not consistent with little_h"
+
+
+def raise_warning(warning, verbose=True):
+    """Warning function."""
+    if verbose:
+        print('Warning: {}'.format(warning))
```

Comment:
i don't understand the point of this function... why not use the `logging` module? Or just do a `from warnings import warn`?

Reply:
This was to have the `verbose` option, that is if you run the `UVWindow` module with `verbose=False`, the messages are not printed to screen. But maybe it is not that useful - I switched to `warnings.warn`

Reply:
yeah, if you really want to have the explicit "verbose" option, I'd just do:

```python
if verbose:
    warnings.warn("message")
```
in your main code, rather than calling another function to do it. But really, if you use `warnings.warn` (or even better, `logger.warn`) then the user can switch on/off the warnings (or other output) without passing that parameter around at all.

Reply:
Okay, that makes sense, thanks. None of the other `hera_pspec` modules use a logger so I guess I'll keep to `warnings.warn`